z3.cabal

```diff
@@ -62,7 +62,8 @@ Library
     -- Ban to mtl-2.1: modify in MonadState causes <<loop>> in mtl-2.1
     Build-depends:       base >= 4.5 && <5,
                          containers,
-                         transformers >= 0.2
+                         transformers >= 0.2,
+                         concurrent-extra
```

Comment:
I'm a bit concerned about depending on `concurrent-extra`... the library  hasn't been updated in three years. Although I have not found a better alternative yet.

Reply:
The code that we need from the library is not much so I would be inclined to just have our own implementation of `RLock`, we can mostly copy-paste (ACK-ing where the code came from). We just need the `with` function. What do you think?

Reply:
Indeed we could do our own locking using a plain `MVar`, but here I would prefer to use the existing package. In the previous case https://github.com/IagoAbal/haskell-z3/pull/59#pullrequestreview-604658308 of `Control.Monad.Extra( whenMaybe )` I agree that the dependency was not worth a simple one line function. But here the complexity is worth a dependency in my opinion. Personally, I think it is better to re-use an existing package and fix any potential issues there than reproducing the code here. And if the library is no longer maintained we can always fork it.

I am not very familiar with the Haskell/hackage ecosystem yet so I default to my general opinion of small modules that do few things, but I am open to hearing more reasons for avoiding the dependency.

Reply:
I tend to be very careful with adding dependencies. Third-party not-very-well-maintained libraries just add risk and can make maintenance more cumbersome. Looking at https://github.com/basvandijk/concurrent-extra/issues I would say the project is orphan, and I think it's not much code (~100 LOC) involved in implementing an `RLock` anyways. But maybe what we can do instead is to use `MVar`s directly and only serialize the finalizers, instead of trying to make the entire library thread-safe. In this case an `MVar ()` would be enough.

Reply:
Okay, that is also fine with me. The project does look unmaintained. I will port the necessary code and remove the dependency :+1: